Only return domains that are eligible to link downstream

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SS-173

For the GAed version of Linked Projects, we intend to only allow domains under the same account to be linked, and require users to be admins as well. This change only fetches potential downstream domains to display that the current user is an admin in (and checks that the user is an admin in the upstream domain), as well as checks for the linked projects privilege. This change also leaves the existing feature flagged version in tact to allow for grandfathered in projects to continue using linked projects as they have been. There are other places I will need to make changes similar to this, but this felt like as good a place as any to start.

I do wonder if in the future we wanted to give other plans access to this privilege, does limiting domains to the same account hinder us at all? I don't have the best understanding of what a typical HQ setup might be for customers.

Also I know there is similar code to fetch domains to copy an app to via the app manager:
https://github.com/dimagi/commcare-hq/blob/7294a095052eaa0771709d7bf16f69549d8847ac/corehq/apps/app_manager/views/view_generic.py#L264-L271
but I don't think we want to make this same change there since that list of domains is shared between creating domain links and copying apps. I think the best future change would be to update the UI for linking an app to be similar to how reports and keywords are linked. That way we can pull two distinct lists that make sense in their respective contexts.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Refactored existing automated tests to cover new behaviors.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
limit GAed version to enterprise and admin users

orangejenny commented:
Haven't read this thoroughly yet, but would it simplify things to only allow new linkages within the same enterprise, while still supporting legacy links that cross account boundaries? (Assuming that's okay from a product/business perspective.)

gherceg committed:
default object manager for DomainLink already excludes deleted links

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains

gherceg commented:
The goal is to keep linked projects functioning _as it was before_ which means we would still have to support creating domain links. This change allows the feature flagged version to function as it did before, and only changes behavior for those with the linked projects privilege enabled. I will be updating that privilege to not be on enterprise accounts by default to avoid rolling this out to USH earlier than anticipated.

orangejenny commented on corehq/apps/linked_domain/tests/test_dbaccessors.py:
```diff
@@ -12,52 +24,154 @@ class TestGetAvailableDomainsToLink(TestCase):
+                                   SoftwarePlanEdition.ENTERPRISE)
 
         cls.user = WebUser.create(
-            domain=cls.test_domain_obj1.name,
```
Naming: I'd keep the "_obj" since it's already being used, see [here](https://github.com/dimagi/commcare-hq/blob/42173e1511bbb9d86536273687219455c579dc06/corehq/apps/domain/models.py#L293-L300).

gherceg replied:
Good point. I made that change and used `create_domain` instead of `arbitrary_domain` after our book club conversation this morning.

gherceg committed:
address feedback

- include obj in domain vars
- use explicit domain names

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains

gherceg committed:
replace linked_projects with release_management

gherceg committed:
refactor tests to check for non-enterprise privilege

NOTE: this will be rolled out to all enterprise accounts in the near future
so will have to change back soon

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains
